arbor/include/arbor/communication/remote.hpp

```diff
@@ -0,0 +1,182 @@
+#pragma once
+
+#include <cstring>
+#include <iostream>
+#include <memory>
+#include <cstdint>
+#include <exception>
+#include <variant>
+
+#include <mpi.h>
+
+namespace arb {
+namespace remote {
+
+// Magic protocol tag
+constexpr std::uint8_t ARB_REMOTE_MAGIC = 0xAB;
+
+// Remote protocol version
+constexpr std::uint8_t ARB_REMOTE_VERSION_MAJOR = 0x00;
+constexpr std::uint8_t ARB_REMOTE_VERSION_MINOR = 0x01;
+constexpr std::uint8_t ARB_REMOTE_VERSION_PATCH = 0x00;
+
+// Message buffer length
+constexpr std::size_t ARB_REMOTE_MESSAGE_LENGTH = 1024;
+
+// Who sends the control message?
+constexpr int ARB_REMOTE_ROOT = 0;
+
+// Messages
+struct msg_null {
+    static constexpr std::uint8_t tag = 0x00;
+    std::uint8_t null = 0x0;
+};
+
+struct msg_abort {
+    static constexpr std::uint8_t tag = 0x01;
+    char reason[512];
+};
+
+struct msg_epoch {
+    static constexpr std::uint8_t tag = 0x02;
+    float t_start;
+    float t_end;
+};
+
+struct msg_done {
+    static constexpr std::uint8_t tag = 0x03;
+    float time = 0.0f;
+};
+
+using ctrl_message = std::variant<msg_null,
+                                  msg_abort,
+                                  msg_epoch,
+                                  msg_done>;
+
+// Exceptions
+struct remote_error: std::runtime_error {
+    remote_error(const std::string& msg): std::runtime_error{msg} {}
+};
+
+struct unexpected_version: remote_error {
+    unexpected_version(): remote_error{"Arbor remote: Magic or Version mismatch."} {}
+};
+
+struct unexpected_message: remote_error {
+    unexpected_message(): remote_error{"Arbor remote: Received unknown tag."} {}
+};
+
+struct mpi_error: remote_error {
+    mpi_error(const std::string& where, int rc):
+        remote_error{"MPI failed in " + where + " with error=" + std::to_string(rc)} {}
+};
+
+inline
+void mpi_checked(int rc, const std::string& where) {
+    if (rc != MPI_SUCCESS) throw mpi_error{where, rc};
+}
+
+// Prepare byte buffer for sending.
+inline
+std::vector<char> make_send_buffer(const ctrl_message& msg, std::vector<char>& send) {
+    std::size_t ptr = 0;
```

Comment:
check for size of `send` is missing

Reply:
This is an internal helper only. We sadly cannot hide these, so I'll inline them.